readthedocs/proxito/views/serve.py

```diff
@@ -623,6 +578,194 @@ def _register_broken_link(self, project, version, path, full_path):
+                    storage_root_path, tryfile
+                )
+                if build_media_storage.exists(storage_filename_path):
+                    log.info(
+                        "Serving custom 404.html page.",
+                        version_slug_404=version_slug_404,
+                        storage_filename_path=storage_filename_path,
+                    )
+                    resp = HttpResponse(
+                        build_media_storage.open(storage_filename_path).read()
+                    )
+                    resp.status_code = 404
+                    return resp
+        return None
+
+    def _get_index_file_redirect(self, request, project, version, filename, full_path):
+        """Check if a file is a directory and redirect to its index/README file."""
+        storage_root_path = project.get_storage_path(
+            type_="html",
+            version_slug=version.slug,
+            include_file=False,
+            version_type=self.version_type,
+        )
+
+        # First, check for dirhtml with slash
+        for tryfile in ("index.html", "README.html"):
+            storage_filename_path = build_media_storage.join(
+                storage_root_path,
+                f"{filename}/{tryfile}".lstrip("/"),
+            )
+            log.debug("Trying index filename.")
+            if build_media_storage.exists(storage_filename_path):
+                log.info("Redirecting to index file.", tryfile=tryfile)
+                # Use urlparse so that we maintain GET args in our redirect
+                parts = urlparse(full_path)
+                if tryfile == "README.html":
+                    new_path = parts.path.rstrip("/") + f"/{tryfile}"
+                else:
+                    new_path = parts.path.rstrip("/") + "/"
+
+                # `full_path` doesn't include query params.`
+                query = urlparse(request.get_full_path()).query
+                redirect_url = parts._replace(
+                    path=new_path,
+                    query=query,
+                ).geturl()
+
+                # TODO: decide if we need to check for infinite redirect here
+                # (from URL == to URL)
+                return HttpResponseRedirect(redirect_url)
+
+        return None
+
+    def get_using_unresolver(self, request, path):
+        """
+        404 handler using the new proxito implementation.
+
+        This is basically a copy of the get() method, but adapted to make use
+        of the unresolver to extract the current project, version, and file.
+        """
+        unresolved_domain = request.unresolved_domain
+        # We force all storage calls to use the external versions storage,
+        # since we are serving an external version.
+        if unresolved_domain.is_from_external_domain:
+            self.version_type = EXTERNAL
+
+        project = None
+        version = None
+        filename = None
+        lang_slug = None
+        version_slug = None
```

Comment:
This is ready for review... but I don't like that we are defining lots of variables that are kind of related, this is kind of a copy/paste of the old implementation, but maybe there is a better way...

Reply:
They are mainly needed for custom redirects, maybe the answer is there... if we can simplify that.

Reply:
And, yeah, we should be able to remove `version_slug` and `lang_slug` with https://github.com/readthedocs/readthedocs.org/issues/10112 :)

Reply:
yeah, I don't like too much the one-line assignment too much because it's harder to read the name of the variables. But that's a personal preference 😄